src/include/catalog/catalog_sql_table.h

```diff
@@ -0,0 +1,718 @@
+  RowIterator begin(transaction::TransactionContext *txn) {
+    // initialize all the internal state of the iterator, via constructor
+    // return the first row pointer (if there is one)
+    return RowIterator(txn, this, true);
+  }
+
+  /**
+   * Returns the end iterator
+   * @param txn transaction
+   * @return end iterator
+   */
+  RowIterator end(transaction::TransactionContext *txn) { return RowIterator(txn, this, false); }
+
+  /**
+   * Row iterator for SqlTable
+   */
+  class RowIterator {
+   public:
+    /**
+     * Iterator dereference.
+     */
+    storage::ProjectedColumns &operator*() { return *proj_col_bufp; }
+    /**
+     * Arrow operator.
+     */
+    storage::ProjectedColumns *operator->() { return proj_col_bufp; }
+
+    /**
+     * pre-fix increment only.
+     */
+    RowIterator &operator++() {
+      if (dtsi_ == tblrw_->GetSqlTable()->end()) {
+        // no more tuples. Return end()
+        proj_col_bufp = nullptr;
+        return *this;
+      }
+      while (dtsi_ != tblrw_->GetSqlTable()->end()) {
+        tblrw_->GetSqlTable()->Scan(txn_, &dtsi_, proj_col_bufp);
+        if (proj_col_bufp->NumTuples() != 0) {
+          break;
+        }
+      }
+      if ((dtsi_ == tblrw_->GetSqlTable()->end()) && (proj_col_bufp->NumTuples() == 0)) {
+        // no more tuples
+        proj_col_bufp = nullptr;
+        return *this;
+      }
+      auto layout = tblrw_->GetLayoutP();
+      storage::ProjectedColumns::RowView row_view = proj_col_bufp->InterpretAsRow(*layout, 0);
+      auto ret_vec = tblrw_->ColToValueVec(row_view);
+      return *this;
+    }
+
+    /**
+     * Equals operator.
+     */
+    bool operator==(const RowIterator &other) const { return proj_col_bufp == other.proj_col_bufp; }
+
+    /**
+     * Not equals operator.
+     */
+    bool operator!=(const RowIterator &other) const { return !this->operator==(other); }
+
+    /**
+     * Constructor
+     * @param txn transaction
+     * @param tblrw SqlTableRw
+     * @param begin whether a begin operator is contructed.
+     */
+    RowIterator(transaction::TransactionContext *txn, SqlTableRW *tblrw, bool begin)
+        : txn_(txn), tblrw_(tblrw), buffer_(nullptr), dtsi_(tblrw->GetSqlTable()->begin()), layout_(nullptr) {
+      if (!begin) {
+        // constructing end
+        proj_col_bufp = nullptr;
+        return;
+      }
+      layout_ = tblrw_->GetLayoutP();
+      all_cols = StorageTestUtil::ProjectionListAllColumns(*layout_);
+      auto col_initer = new storage::ProjectedColumnsInitializer(*layout_, all_cols, 1);
+      buffer_ = common::AllocationUtil::AllocateAligned(col_initer->ProjectedColumnsSize());
+      proj_col_bufp = col_initer->Initialize(buffer_);
+
+      while (dtsi_ != tblrw_->GetSqlTable()->end()) {
+        tblrw_->GetSqlTable()->Scan(txn, &dtsi_, proj_col_bufp);
+        if (proj_col_bufp->NumTuples() == 0) {
+          continue;
+        }
+        break;
+      }
+      if (dtsi_ == tblrw_->GetSqlTable()->end()) {
```

Comment:
There seems to be a bug here. We should first check if NumTuples() is 0 before ending prematurely.

Reply:
I'm not clear on where you think that check should be... can you clarify please?

Reply:
The check should in the if statement. I come across this error when there is only one tuple in the table. Here is what I see:
1. SqlTableRW::begin() is called. Because there is only one tuple, dsti_ will reach the end of the table after one call to Scan.
2. Without this check, proj_col_bufp will be set to nullptr since dtsi_ is at the end of the table.
3. Now begin() == end() even though the table is not empty.
Am I maybe using this object incorrectly?

Reply:
Nice catch. Your logic is sound, but I have not been able to recreate this bug. I've only had a small amount of time to try. Nevertheless, I added a check, so that the iterator will not prematurely set the termination condition.